File: python/learn_invariance.py
from numpy import *
from tools import *


def learn_invariance(weight, param, common):
    '''Timothee learning algorithm'''
    
    #Set random seed for test run
    if params['test_mode']:
        random.mtrand.seed(5489)

    #when true, reinforce when previous input and current output correlate
    #(eg Einhauser 2002)
    correl_prev_input_curr_output = False 
    geom_factor = (param['learning_rate_f'] \
                   / param['learning_rate_i']) ** \
        (1.0 / floor(param['n_iter'] / 1000.0))
    param['a_pos'] = param['learning_rate_i']
    param['a_neg'] = -param['a_pos'] / (param['n_feat'] * \
                                        prod(param['rf_size'])) * \
        param['correct_ratio']
   
    if weight == []:
        weight = zeros((param['rf_size'][0], param['rf_size'][1],
                       param['n_feat'], param['n']), float64)
        for n_ind in range(param['n']):
            weight[:, :, :, n_ind] = get_initial_weight(param['rf_size'],
                                                        param['n_feat'])
            
        #init middle layer
        if param['input_nu'] < inf:
            tr_input = 1.0 / param['input_nu'] * ones((param['rf_size'][0], 
                                                      param['rf_size'][1],
                                                      param['n_feat']), float64)
        else:
            tr_input = ones((param['rf_size'][0], param['rf_size'][1],
                            param['n_feat']), float64)

        if param['output_nu'] < inf:
            tr_output = 1.0 / param['output_nu'] * ones(param['n'], float64)
        else:
            tr_output = ones(param['n'], float64)

        input_threshold = zeros((param['rf_size'][0], param['rf_size'][1],
                                param['n_feat']), float64)
        output_threshold = ones(param['n'], float64)

        previous_winner_pair = -1.0 * ones((4), float64)

        # reporting
        n_firing_output = zeros((param['n']), float64)
        n_firing_input = zeros((param['rf_size'][0], param['rf_size'][1],
                               param['n_feat']), float64)
        n_above_thr = zeros((param['n_iter']), float64) 

        #init evol
        evol = zeros((param['n'], floor(param['n_iter'] / 100)), float64)

    print 'Learning invariances...'
    offset = floor(random.random() * param['n_iter'])

    
    for i_ind in range(param['n_iter']):
        
        effective = False
        #current input
        gross_input, common = get_s1_map(i_ind + offset, common)
    
        #update traces
        tr_input = gross_input / param['input_nu'] + \
        (1.0 - 1.0 / param['input_nu']) * tr_input
        #normalization by traces
        input_var = gross_input / tr_input
        n_above_thr[i_ind]  = (input_var > input_threshold).sum()
        
        #get max input
        input_max, idx = multi_dimensional_max(input_var)
        input_winner_i = idx[0]
        input_winner_j = idx[1]
        input_winner_f = idx[2]
        gross_input_max = gross_input[idx[0], idx[1], idx[2]]

        #current output
        gross_output = zeros(param['n'], float64)
        for n_ind in range(param['n']):
            if param['use_soft_max']:
                #softmax
                w = weight[:, :, :, n_ind]
                gross_output[n_ind] = sof_max(w, input_var, param['p'],
                                              param['q'], param['r'])
            else:
                gross_output[n_ind] = (weight[:, :, :, n_ind] * input_var).max()


        #update traces
        if i_ind == 0:
            tr_output = gross_output
        else:
            tr_output = gross_output / param['output_nu'] + \
            (1.0 - 1.0 / param['output_nu']) * tr_output

        #normalization by traces
        output = zeros(param['n'], float64)
        for n_ind in range(param['n']):
            output[n_ind] = gross_output[n_ind] / tr_output[n_ind]

        output_max = tr_output.max()
        output_winner = tr_output.argmax()

        #correlation between previous input and current output
        #(like Einhauser 2002)
        if correl_prev_input_curr_output: 
            #Learning
            if i_ind > 0 and previous_input_max > \
            input_threshold[previous_input_winner_i, previous_input_winner_j,
            previous_input_winner_f]:
                #update threshold
                input_threshold[previous_input_winner_i,
                previous_input_winner_j, previous_input_winner_f] = \
                param['thr_coef'] * previpus_input_max
                #update wiehgts: stpre winner-winner weight
                win_weight = weight[previous_input_winner_i,
                previous_input_winner_j, previous_intput_f, output_winner]
                #depress everyone
                weight[:, :, :, output_winner] = \
                weight[:, :, :, output_winner] + param['a_neg'] * \
                weight[:, :, :, output_winner] * (1-weight[:, :, :,
                                                  output_winner])

                #reproting
                n_firing_input[previous_input_winner_i, previous_winner_j,
                previous_input_winner_f] = \
                n_firing_input[previous_input_winner_i, previous_input_winner_k,
                previous_winner_f] + 1
                n_firing_output[output_winner] = \
                n_firing_output[output_winner] + 1
        else: #correlation between previous output and current input
            #learning
            if (i_ind > 0) and (input_max > input_threshold[input_winner_i,
                                input_winner_j, input_winner_f]):
                #update the
                input_threshold[input_winner_i, input_winner_j,
                input_winner_f] = param['thr_coef'] * input_max
                #store winner weight
                win_weight = weight[input_winner_i, input_winner_j,
                input_winner_f, previous_output_winner]
                #depress everyone
                weight[:, :, :, previous_output_winner] = \
                weight[:, :, :, previous_output_winner] + param['a_neg'] * \
                weight[:, :, :, previous_output_winner] * \
                (1.0-weight[:, :, :, previous_output_winner])
                #potentiate synapse between winners
                weight[input_winner_i, input_winner_j, input_winner_f,
                previous_output_winner] = win_weight + param['a_pos'] * \
                win_weight * (1.0 - win_weight)
                # reporting
                n_firing_input[input_winner_i, input_winner_j,
                input_winner_f] = n_firing_input[input_winner_i, input_winner_j,
                input_winner_f] + 1
                n_firing_output[previous_output_winner] = \
                n_firing_output[previous_output_winner] + 1


        weight = maximum(weight, 0)
        weight = minimum(weight, 1)

        
        if (mod(i_ind + 1, 100) == 0):
            evol[:, int((i_ind + 1) / 100) - 1] = weight.sum()

        #threshold decay
        input_threshold = input_threshold * (1.0 - param['thr_decay'])
        #previous := current
        previous_input_max = input_max
        previous_input_winner_i = input_winner_i
        previpus_input_winner_j = input_winner_j
        previous_input_winner_f = input_winner_f
        previous_output_winner = output_winner
        previous_output_max = output_max
        previous_gross_input_max = gross_input_max

        if mod(i_ind + 1, 1000) == 0:
            param['a_pos'] = param['a_pos'] * geom_factor
            param['a_neg'] = param['a_neg'] * geom_factor
        
        if mod(i_ind, round(param['n_iter'] / 100)) == 0:
            print '.'

    print '\ntr ' + str(tr_output)
    print 'thr ' + str(output_threshold)
    print 'out' + str(output)

    common['weight'] = weight 
    common['evol'] = evol 
    common['thr'] = input_threshold 
    common['n_firing_input'] = n_firing_input
    common['n_firing_output'] = n_firing_output 
    common['n_above_thr'] = n_above_thr
    return common


def get_initial_weight(patch_size, n_feat):

    weight = 0.75 * (1.0 + 0.0 * (random.random((patch_size[0], patch_size[1], n_feat))- 0.5))
    return weight 

File: matlab/learnInvariance.m
function [weight, evol, inputThreshold, nFiringInput, nFiringOutput, nAboveThr]=learnInvariance(getMap,weight,PARAM)
% Timothee Masquelier timothee.masquelier@alum.mit.edu August 2006
% see Einhauser 2002, Spratling 2005
rand('twister',5489);

CORREL_PREV_INPUT_CURR_OUTPUT = false;  % when true, reinforce when previous input and current output correlate (eg Einhauser 2002)
% when false, reinforce when previous output and current input correlate(eg Spratling 2005)
% false much more robust (Tim 06/2007)
% global PARAM
% global COMMON

% PARAM.nu = 100;
% PARAM.thresoldAM = .05;
% PARAM.thrDecay = 2^-11; % optimal around 2^-11
geomFactor = ( PARAM.learningRate_f / PARAM.learningRate_i )^(1/floor(PARAM.nIter/1000));
PARAM.a_pos = PARAM.learningRate_i;
PARAM.a_neg = -PARAM.a_pos/(PARAM.nFeat*prod(PARAM.RFSize))*PARAM.correctRatio;

if isempty(weight)
    for n=1:PARAM.n
        % format: i x j x feat x node
        weight(:,:,:,n) = getInitialWeight(PARAM.RFSize,PARAM.nFeat);
    end
    % init middle layer
    %     trAM = 1/PARAM.nu*ones([PARAM.RFSize PARAM.nFeat]);% format: i x j x feat
    if PARAM.inputNu < Inf
        trInput = 1/PARAM.inputNu*ones([PARAM.RFSize PARAM.nFeat]);% format: i x j x feat
    else
        trInput = ones([PARAM.RFSize PARAM.nFeat]);% format: i x j x feat;
    end
    if PARAM.outputNu < Inf
        trOutput = 1/PARAM.outputNu*ones(1,PARAM.n);
    else
        trOutput = ones(1,PARAM.n);
    end
    %     thresholdAM = .00*ones([PARAM.RFSize PARAM.nFeat]);% format: i x j x feat
    %     inputThreshold = 0.055*ones([PARAM.RFSize PARAM.nFeat]);% format: i x j x feat
    inputThreshold = zeros([PARAM.RFSize PARAM.nFeat]);% format: i x j x feat
    outputThreshold = ones(1,PARAM.n);

    previousWinnerPair = -1*ones(1,4); % format: inputWinnerI winnerInputJ winnerInputF previousOutputWinner

    %     previousMaxAM = -1;
    %     previousWinnerMI = -1; % has no impact
    %     previousWinnerMJ = -1; % has no impact
    %     previousWinnerMF = -1; % has no impact

    % init top layer
    %     trAT = 1/PARAM.nu*ones(1,PARAM.n);
    %     previousMaxAT = -1;
    %     previousWinnerT = -1;

    %reporting
    nFiringOutput = zeros(1,PARAM.n);
    nFiringInput = zeros([PARAM.RFSize PARAM.nFeat]);
    nAboveThr = zeros(1,PARAM.nIter);

    % init evol
    evol = zeros(1,PARAM.n,floor(PARAM.nIter/100));
end
disp('Learning invariances...')
tic
offset = floor(rand*PARAM.nIter);
% disp(int2str(offset))

for i=1:PARAM.nIter

    effective = false;

    % CURRENT INPUT
    grossInput = feval(getMap,i+offset); % format: i x j x feat

    % update traces
    %     if ~PARAM.trInputEffective || effective % only compute average with effective stim
    %         trInput = input / PARAM.inputNu + (1-1/PARAM.inputNu) * trInput;
    trInput = grossInput / PARAM.inputNu + (1-1/PARAM.inputNu) * trInput;
    %     end

    % normalization by the trace
    input = grossInput ./ trInput; % format: i x j x feat

    % n above thr
    nAboveThr(i) = sum(sum(sum(input>inputThreshold)));
    %     fprintf(1,'%d ', nAboveThr(i));

    % get max input
    [ inputMax idx ] =  multiDimensionalMax(input);
    inputWinnerI = idx(1);
    inputWinnerJ = idx(2);
    inputWinnerF = idx(3);
    grossInputMax = grossInput(idx(1),idx(2),idx(3));

    % CURRENT OUTPUT
    
    for n=1:PARAM.n
        if PARAM.useSoftMax
            % softmax
            w = weight(:,:,:,n);
            grossOutput(n) = softmax(w(:),input(:),PARAM.p,PARAM.q,PARAM.r);
        else
            grossOutput(n) = max(max(max( weight(:,:,:,n).*input )));
        end
    end

    % update traces
    %     if ~PARAM.trOutputEffective || effective % only compute average with effective stim
    %         trOutput = output / PARAM.outputNu + (1-1/PARAM.outputNu) * trOutput;
    if i==1
        trOutput = grossOutput;
    else
        trOutput = grossOutput / PARAM.outputNu + (1-1/PARAM.outputNu) * trOutput;
    end
    %     end

    % NOT USED ANYMORE update output thr
%     outputThreshold(previousOutputWinner) = outputThreshold(previousOutputWinner)*(previousOutputMax/PARAM.outputNu + (1-1/PARAM.outputNu) ) ;

    % normalization by the trace
    for n=1:PARAM.n
        output(n) = grossOutput(n) / trOutput(n);
    end

    % find winner in top layer
%     [outputMax outputWinner] = max(output);
            [outputMax outputWinner] = max(trOutput);

    if CORREL_PREV_INPUT_CURR_OUTPUT % correlation between previous input and current output (like Einhauser 2002)

        % learning
        %         if i>1 && previousGrossInputMax > inputThreshold(previousInputWinnerI,previousInputWinnerJ,previousInputWinnerF)
        if i>1 && previousInputMax > inputThreshold(previousInputWinnerI,previousInputWinnerJ,previousInputWinnerF)
            %             effective = true;
            % update thr
            inputThreshold(previousInputWinnerI,previousInputWinnerJ,previousInputWinnerF) = ...
                PARAM.thrCoef*previousInputMax;
            %                 PARAM.thrCoef*previousGrossInputMax;
            % update weights:
            % store winner-winner weight
            winWeight = weight(previousInputWinnerI,previousInputWinnerJ,previousInputWinnerF,outputWinner);
            
            % depress everyone
            weight(:,:,:,outputWinner) = weight(:,:,:,outputWinner) + PARAM.a_neg*weight(:,:,:,outputWinner).*(1-weight(:,:,:,outputWinner));
            % potentiate synapse between winners
            weight(previousInputWinnerI,previousInputWinnerJ,previousInputWinnerF,outputWinner) = winWeight + PARAM.a_pos*winWeight*(1-winWeight);
            
%             % Einhauser 2002
%             % depress everyone
%             weight(:,:,:,outputWinner) = weight(:,:,:,outputWinner) - PARAM.a_pos/256*weight(:,:,:,outputWinner);
%             % potentiate synapse between winners
%             weight(previousInputWinnerI,previousInputWinnerJ,previousInputWinnerF,outputWinner) = winWeight + PARAM.a_pos/256*(1-winWeight);

            
            % reporting
            nFiringInput(previousInputWinnerI,previousInputWinnerJ,previousInputWinnerF) = ...
                nFiringInput(previousInputWinnerI,previousInputWinnerJ,previousInputWinnerF)+ 1;
            nFiringOutput(outputWinner) = nFiringOutput(outputWinner) + 1;
        end % learning

    else % correlation between previous output and current input

        % learning
        if i>1 && inputMax > inputThreshold(inputWinnerI,inputWinnerJ,inputWinnerF)
            %         if i>1 && grossInputMax > inputThreshold(inputWinnerI,inputWinnerJ,inputWinnerF)
            %             effective = true;
            % update thr
            inputThreshold(inputWinnerI,inputWinnerJ,inputWinnerF) = PARAM.thrCoef*inputMax;
            %             inputThreshold(inputWinnerI,inputWinnerJ,inputWinnerF) = PARAM.thrCoef*grossInputMax;
            %             inputThreshold(inputWinnerI,inputWinnerJ,inputWinnerF) = inputMax * inputThreshold(inputWinnerI,inputWinnerJ,inputWinnerF);
            %             inputThreshold(inputWinnerI,inputWinnerJ,inputWinnerF) = inputMax;
            % update weights:
            %             if sum( [inputWinnerI inputWinnerJ inputWinnerF previousOutputWinner] ~= previousWinnerPair ) ~= 0 % different pair of winners
            % store winner-winner weight
            winWeight = weight(inputWinnerI,inputWinnerJ,inputWinnerF,previousOutputWinner);

            % depress everyone
            weight(:,:,:,previousOutputWinner) = weight(:,:,:,previousOutputWinner) + PARAM.a_neg*weight(:,:,:,previousOutputWinner).*(1-weight(:,:,:,previousOutputWinner));
            % potentiate synapse between winners
            weight(inputWinnerI,inputWinnerJ,inputWinnerF,previousOutputWinner) = winWeight + PARAM.a_pos*winWeight*(1-winWeight);


%             % Foldiak 1991
%             weight(:,:,:,previousOutputWinner) = weight(:,:,:,previousOutputWinner) + PARAM.a_pos * 2^-8 * ...
%                         (input-weight(:,:,:,previousOutputWinner));
            
            % reporting
            nFiringInput(inputWinnerI,inputWinnerJ,inputWinnerF) = ...
                nFiringInput(inputWinnerI,inputWinnerJ,inputWinnerF)+ 1;
            nFiringOutput(previousOutputWinner) = nFiringOutput(previousOutputWinner) + 1;

            % store winner for future comparison
            previousWinnerPair = [inputWinnerI inputWinnerJ inputWinnerF previousOutputWinner];
            %             end % different winner
        end % learning


    end % switch between the 2 possible correlations
    
    weight = max(weight,0); % just in case
    weight = min(weight,1); % just in case

    %     for ti=1:size(grossInput(1))% i x j x feat
    %         for tj=1:size(grossInput(2))
    %             for tf=1:size(grossInput(3))
    %                 if input(ti,tj,tf)>inputThreshold(ti,tj,tf)
    % %                 if grossInput(ti,tj,tf)>inputThreshold(ti,tj,tf)
    %                     inputThreshold(ti,tj,tf) = PARAM.thrCoef*input(ti,tj,tf);
    % %                     inputThreshold(ti,tj,tf) = PARAM.thrCoef*grossInput(ti,tj,tf);
    %                 end
    %             end
    %         end
    %     end

    if mod(i,100)==0
        %         evol(:,:,i/100) = reshape(sum(sum(weight,1),2),[PARAM.nFeat PARAM.n]); % format: feat x node x iter
        evol(1,:,i/100) = sum(sum(sum(weight,1),2),3); % format: 1 x node x iter
    end

    %     % thr decay
    inputThreshold = inputThreshold*(1-PARAM.thrDecay);
    
    % previous := current
    previousInputMax = inputMax;
    previousInputWinnerI = inputWinnerI;
    previousInputWinnerJ = inputWinnerJ;
    previousInputWinnerF = inputWinnerF;
    previousOutputWinner = outputWinner;
    previousOutputMax = outputMax;
    previousGrossInputMax = grossInputMax;

    if mod(i,1000)==0
        PARAM.a_pos = PARAM.a_pos * geomFactor;
        PARAM.a_neg = PARAM.a_neg * geomFactor;
    end
    if mod(i,round(PARAM.nIter/100))==0
        fprintf(1,'.');
    end
end

fprintf(1,'\n');
toc
%debug
disp(['tr ' num2str(trOutput)]);
disp(['thr ' num2str(outputThreshold)]);
disp(['out ' num2str(output)]);

function weight = getInitialWeight(patchSize,nFeat)
% weight = 1 + .01*rand([patchSize nFeat]);
% weight = max(0,weight);
% weight = weight / sum(weight(:));
weight = .75*( 1 + .0 * (rand([patchSize nFeat])-.5) );
% weight = .5 * ones([patchSize nFeat]);

% function input = tmp(patchSize,nFeat)
% input = rand(patchSize,patchSize,nFeat);
